### PR TITLE
Iow 289

### PR DESCRIPTION
**Pull request by orosz-usgs**
Before making a pull request
----------------------------

- [ ] Make sure all tests run
- [ ] Update the changelog appropriately

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java [hunk 1]
```diff
@@ -76,6 +79,54 @@ public String getOgcCollection(@RequestParam(value = "f", required = false, defa
...
+			)
+	@GetMapping(value = "collections/{collectionId}/items", produces = MediaType.APPLICATION_JSON_VALUE)
+	public String getOgcCollectionFeatures(
+			@RequestParam(value = "f", required = false, defaultValue = "json") String mimeType,
+			@RequestParam(value = "limit", required = false, defaultValue = "${query.limit.max}") Integer limit,
+			@RequestParam(value = "startIndex", required = false, defaultValue = "0") int startIndex,
+			@RequestParam(value = "bbox", required = false) List<Double> bbox,
+			@PathVariable(value = "collectionId") String collectionId, HttpServletResponse response) {
+
+		String rtn = getOgcCollection(mimeType, collectionId, response);
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java]: Why is this being done? (calling the collections endpoint)

Same question in getOgcCollectionFeature

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java [hunk 2]
```diff
@@ -76,6 +79,54 @@ public String getOgcCollection(@RequestParam(value = "f", required = false, defa
...
+		if (response.getStatus() == HttpServletResponse.SC_OK) {
+			Map<String, Object> params = collectionsParams.getParameters(collectionId);
+			int count = collectionsDao.getCollectionFeatureCount(params);
+
+			Integer limitParam = limit;
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java]: Is there any reason to continue if the count is zero (or less than the startIndex)

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java [hunk 3]
```diff
@@ -76,6 +79,54 @@ public String getOgcCollection(@RequestParam(value = "f", required = false, defa
...
+			if (limitParam == null) {
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java]: is this limitParam logic redundant with the code in CollectionParams?
same with the other paging related parameters.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 4]
```diff
@@ -36,4 +42,51 @@
...
 
+	public String getServerUrl() {
+		return serverUrl;
+	}
+
+	public int getMaxLimit() {
+		return maxLimit;
+	}
+
+	public void setLimit(Map<String, Object> params, int limit) {
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Please do not mutate maps this way.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 5]
```diff
@@ -36,4 +42,51 @@
...
+		params.put("limit",limit);
+	}
+
+	public void setStartIndex(Map<String, Object> params, int startIndex) {
+		params.put("startIndex",startIndex);
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Please do not mutate maps this way.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 6]
```diff
@@ -36,4 +42,51 @@
...
+
+	/*
+	 * a bounding box with four numbers:
+     * Lower left corner,  longitude
+     * Lower left corner,  latitude
+     * Upper right corner, longitude
+     * Upper right corner, latitude
+	 */
+	public void setBbox(Map<String, Object> params, List<Double> bbox) {
+		if (bbox != null && bbox.size() == 4) {
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Please do not mutate maps this way.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 7]
```diff
@@ -36,4 +42,51 @@
...
+			params.put("pointLowLeft", String.format("Point(%f %f)", bbox.get(0), bbox.get(1)));
+			params.put("pointUpRight", String.format("Point(%f %f)", bbox.get(2), bbox.get(3)));
+		}
+	}
+
+	public void setPagingParams(Map<String, Object> params, int count) {
+		int startIndex = params.get("startIndex") == null ? START_INDEX_DEFAULT : (Integer) params.get("startIndex");
```
Comment (dsteinich) [inline src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Please do not mutate maps this way.

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -40,3 +40,7 @@ collections:
   monitoring-locations:
     title: NWIS Monitoring Locations
     description: USGS water monitoring locations managed in the National Water Information System
+
+query:
+  limit:
+    max: 10000
```
Comment (dsteinich) [inline src/main/resources/application.yml]: This could just be a constant in the java code. If the goal is to make it configuration, it should be a substitution here with an environment variable to populate it.

File: src/main/resources/mybatis/features.xml [hunk 9]
```diff
@@ -119,4 +218,9 @@
         ) as feature
     </select>
 
+    <select id="getCollectionFeatureCount" parameterType="map" resultType="int">
+      select count(*)
```
Comment (dsteinich) [inline src/main/resources/mybatis/features.xml]: This could be a resource hog.

File: src/main/resources/mybatis/features.xml [hunk 10]
```diff
@@ -36,20 +36,35 @@
 
      <sql id="selectFeaturesQuery">
           (select
-              'monitoring-locations' collection_id, #{monLocTitle,jdbcType=VARCHAR} collectionTitle,
+              'monitoring-locations' collection_id, #{monLocTitle,jdbcType=VARCHAR} collection_title,
```
Comment (dsteinich) [inline src/main/resources/mybatis/features.xml]: For the features query we should always have a single collection id. Use that to our advantage and do not have this union query with the hard to read, interpret, and duplicated login run on UNION

File: src/main/resources/mybatis/features.xml [hunk 11]
```diff
@@ -61,6 +76,35 @@
...
+              limit #{limit}
+           </if>
+          <if test="startIndex != null">
+             offset #{startIndex}
+          </if>
+           )
+          ORDER BY collection_id, feature_id
+     </sql>
+
+     <sql id="selectCollectionTitleQuery">
```
Comment (dsteinich) [inline src/main/resources/mybatis/features.xml]: Why is the collection title logic duplicated (also in the selectFeaturesQuery). Causes confusion and potential double reads of the network table.

File: src/test/java/gov/usgs/wma/waterdata/collections/CollectionsControllerIT.java [hunk 12]
```diff
@@ -93,6 +99,44 @@ public void collectionNotFoundTest() {
 		assertNull(rtn.getBody());
 	}
 
+	@Test
+	public void collectionFeaturesTest() {
+		for (String collectionId : COLLECTION_IDS) {
```
Comment (dsteinich) [inline src/test/java/gov/usgs/wma/waterdata/collections/CollectionsControllerIT.java]: These tests are very hard to follow. Smells like a lot of hard to maintain java code when a simple hamcrest based test would do. see the MonitoringLocationControllerIT.foundTest()

File: src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+	private CollectionParams collectionsParams;
+
+	protected static final List<String> COLLECTION_IDS = List.of("ANC", "monitoring-locations");
+
+	public static final List<String> REL_TYPES = List.of("canonical", "self", "collection", "prev", "next");
+	public static final int MONLOC_COUNT = 110;
+
+	@Test
+	public void featuresDefaultLimitTest() {
+		int maxLimit = collectionsParams.getMaxLimit();
```
Comment (dsteinich) [inline src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java]: Tests too complicated to follow

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 14]
```diff
@@ -36,4 +42,51 @@
...
+		int limit = params.get("limit") == null ? maxLimit : (Integer) params.get("limit");
+
+		if (startIndex > 0) {
+			int prevStartIndex = startIndex - limit;
+			if (prevStartIndex < 0) {
+				prevStartIndex = 0;
+			}
```
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Ed does not see my comment. Testing if this reply makes it available to him.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java [hunk 15]
```diff
@@ -36,4 +42,51 @@
...
+			params.put("prevStartIndex", String.format("&startIndex=%d&limit=%d", prevStartIndex, limit));
+		}
+
+		int nextStartIndex = startIndex + limit;
+		if (nextStartIndex <= count - 1) {
+			params.put("nextStartIndex", String.format("&startIndex=%d&limit=%d", nextStartIndex, limit));
+		}
```
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: Ed does not see my comment. Testing if this reply makes it available to him.

File: src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+				ids.add(featureJson.getString("id"));
+			}
+
+		}
+
+		return ids;
+	}
+
+	private boolean isBetween(double a, double b, double c) {
+		return b > a ? c >= a && c <= b : c >= b && c <= a;
```
[the remark this answers is not in the document]
Comment (duselmann) [reply src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java]: Ed does not see my comment. Testing if this reply makes it available to him.

File: src/main/resources/application.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/resources/application.yml]: This ticket is just for monitoring-locations. I would assume timeseries would have
it own paging.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java [hunk 17]
```diff
@@ -76,6 +79,54 @@ public String getOgcCollection(@RequestParam(value = "f", required = false, defa
...
+				limitParam = collectionsParams.getMaxLimit();
+			}
+			if (limit > collectionsParams.getMaxLimit()) {
+				limitParam = collectionsParams.getMaxLimit();
+			}
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java]: check the latest, this code was refactored

File: src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/test/java/gov/usgs/wma/waterdata/collections/FeaturesParamsIT.java]: After I submitted this, I was thinking this test should also be a hamcrest
file compare. That refactor would eliminate this. This was put it so there would be one
less file to maintain. But the test would still need to be maintained. So not sure
if the trade off would be worth it.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: I think maybe it should be:
int nextStartIndex = Math.min(count, startIndex + limit);
we would still need the if check on line 88
they seem about equivalent.

[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: you would if you do not subtract one from count in the min. if you subtract one then it would be  complete. I would test it of course and make truth tables to make sure they are equal.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionsController.java]: ok

File: src/main/resources/mybatis/features.xml [hunk 18]
```diff
@@ -105,12 +129,68 @@
...
               'rel',  'collection',
               'type', 'application/json',
-               'title', collectionTitle,
+               'title', collection_title,
                'href', #{serverUrl,jdbcType=VARCHAR} || 'collections/' || collection_id || '?f=json'
              )
           ) links
      </sql>
 
+     <select id="getCollectionFeaturesJson" parameterType="map" resultType="String">
```
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/resources/mybatis/features.xml]: It looks like accesslint thinks this is HTML but it is a MyBatis select tag.

**Comment by orosz-usgs** [comment]
yes, this is a server project, accesslint does not need to be on it not be on it.

File: src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/collections/CollectionParams.java]: true, at this point I would prefer to leave as is for now.
